src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -0,0 +1,123 @@
+            String currentWord = userInputWords[i];
+            if (i < userInputWords.length - 1 || userInput.endsWith(" ")) {
+                toggleMilestone(currentWord);
+            } else {
+                filterMilestones(currentWord);
+            }
+        }
+    }
+
+    /**
+     * Gets a milestone that is the ONLY match to the given milestoneQuery
+     * Changes the selection statuses of all milestones in the milestone list,
+     * such that the new given milestone is toggled, and the rest are not selected
+     * @param milestoneQuery
+     */
+    public final void toggleMilestone(String milestoneQuery) {
+        Optional<PickerMilestone> onlyMatchingMilestone = getMatchingMilestoneIfOnlyMatch(milestoneQuery);
+        if (!onlyMatchingMilestone.isPresent()) return;
+        this.currentMilestonesList.stream()
+                .forEach(milestone -> milestone.setSelected(milestone.equals(onlyMatchingMilestone.get())
+                        && !milestone.isSelected()));
+    }
+
+    /**
+     * Gets the current list of milestones
+     * @return
+     */
+    public List<PickerMilestone> getCurrentMilestonesList() {
+        return this.currentMilestonesList;
+    }
+
+    /**
+     * Gets the list of milestones that matches the current query i.e. not faded
+     * @return
+     */
+    public List<PickerMilestone> getMatchingMilestonesList() {
+        return this.currentMilestonesList.stream()
+                .filter(milestone -> !milestone.isFaded())
+                .collect(Collectors.toList());
+    }
+
+    private void filterMilestones(String query) {
+        currentMilestonesList.stream()
+                .forEach(milestone -> {
+                    boolean matchQuery = Utility.containsIgnoreCase(milestone.getTitle(), query);
+                    milestone.setFaded(!matchQuery);
+                });
+
+        if (hasExactlyOneMatchingMilestone(currentMilestonesList, query)) highlightAnyMatchingMilestone();
+    }
+
+    private void highlightAnyMatchingMilestone() {
+        if (!hasMatchingMilestone(this.currentMilestonesList)) return;
+        this.currentMilestonesList.stream()
+                .filter(milestone -> !milestone.isFaded())
+                .findAny()
+                .get()
+                .setHighlighted(true);
+    }
+
+    private boolean hasMatchingMilestone(List<PickerMilestone> milestoneList) {
+        return milestoneList.stream()
+                .filter(milestone -> !milestone.isFaded())
+                .findAny()
+                .isPresent();
+    }
+
+    /**
+     * Gets the name of milestone if it is the only matching milestone to the query
+     * @param query
+     * @return
+     */
+    private Optional<PickerMilestone> getMatchingMilestoneIfOnlyMatch(String query) {
+        if (!hasExactlyOneMatchingMilestone(currentMilestonesList, query)) return Optional.empty();
+
+        return Optional.of(getMatchingMilestone(currentMilestonesList, query));
+    }
```

Comment:
This method is odd. It's better if you have a method that returns all matching milestones and let the caller decide if it has 0 or 1, or more. If you have a method like this, the Optional.empty() could mean either there is no matching milestone or there were multiple matching milestones.